### PR TITLE
Add Keck aperture

### PR DESCRIPTION
**Pull request by vkooten**
This adds Keck aperture ~and lyot stop~ for Keck 2. It solves part of #134 and #153 

Picture of aperture:
keck

### CODE REVIEW

File: hcipy/aperture/realistic.py [hunk 1]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+    with_segment_gaps : TYPE, optional
+        DESCRIPTION. The default is False.
+    gap_padding : TYPE, optional
+        DESCRIPTION. The default is 0.
+    segment_transmissions : TYPE, optional
+        DESCRIPTION. The default is 1.
+    return_header : TYPE, optional
+        DESCRIPTION. The default is False.
+    return_segments : TYPE, optional
+        DESCRIPTION. The default is False.
```
Comment (ehpor) [inline hcipy/aperture/realistic.py:1349]: Put in reasonable descriptions in here.

File: hcipy/aperture/realistic.py [hunk 2]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
 		segments = [segment_with_strut(segment) for segment in segments]
 
 	return func, segments
+
+def make_keck_aperture(normalized=True, with_spiders=False, with_segment_gaps=False, gap_padding=0, segment_transmissions=1, return_header=False, return_segments=False):
+    """
+    This code creates a keck-like aperture matching values used in vanKooten [2021 and 2022] as well as
```
Comment (ehpor) [inline hcipy/aperture/realistic.py:1331]: I'd add in an actual citation for these, so that they get formatted correctly in the compiled documentation.

**Review by ehpor — CHANGES_REQUESTED** [review]
- [ ] There are lots of flake8 errors. You can look at [Azure](https://github.com/ehpor/hcipy/runs/8098734456?check_suite_focus=true) for a full list of them since the annotations only show the first twenty). Or you can run flake8 yourself locally.
- [ ] Lots of whitespace issues. Hcipy uses tabs for indentation (I know... smh; I've been looking for a way to change this without clobbering the whole git history).
- [ ] Add a pytest for the Keck aperture, making sure to check normalization and segmentation.
- [ ] Put a PNG of the baseline aperture (as checked by the Pytest for an easy check for us and documentation).
- [ ] You can also add the Keck aperture (not Lyot stop) to the telescope pupils tutorial. Or I can do that in a separate PR afterwards. Let me know which option you want.

File: hcipy/aperture/realistic.py [hunk 3]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+    with_segment_gaps : TYPE, optional
+        DESCRIPTION. The default is False.
+    gap_padding : TYPE, optional
+        DESCRIPTION. The default is 0.
+    segment_transmissions : TYPE, optional
+        DESCRIPTION. The default is 1.
+    return_header : TYPE, optional
+        DESCRIPTION. The default is False.
+    return_segments : TYPE, optional
+        DESCRIPTION. The default is False.
```
[the remark this answers is not in the document]
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

File: hcipy/aperture/realistic.py [hunk 4]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+    def segment_with_spider(segment):
+        return lambda grid: segment(grid) * spider1(grid) * spider2(grid) * spider3(grid) * spider4(grid)
+    segments = [segment_with_spider(s) for s in segments]
+    contour = make_segmented_aperture(segment, segment_positions)
+
+    def func(grid):
+        ap=contour(grid)
+        co=make_circular_aperture(central_obscuration_diameter)(grid)
+        ap[co==1]=0
+        res = (ap )#* spider1(grid) * spider2(grid) * spider3(grid)* spider4(grid) * spider3(grid)* spider5(grid) * spider6(grid) # * coro(grid)
```
[the remark this answers is not in the document]
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

File: hcipy/aperture/realistic.py [hunk 5]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+    segmentation, segments = segmented_aperture
+    def segment_with_spider(segment):
+        return lambda grid: segment(grid) * spider1(grid) * spider2(grid) * spider3(grid) * spider4(grid)
+    segments = [segment_with_spider(s) for s in segments]
+    contour = make_segmented_aperture(segment, segment_positions)
+
+    def func(grid):
+        ap=contour(grid)
+        co=make_circular_aperture(central_obscuration_diameter)(grid)
+        ap[co==1]=0
```
[the remark this answers is not in the document]
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

File: hcipy/aperture/realistic.py [hunk 6]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+        return Field(res, grid)
+    if return_segments:
+        return func, segments
+    else:
+        return func
+def make_keck_lyot_stop(normalized=True, with_spiders=False, with_segment_gaps=False, gap_padding=0, segment_transmissions=1, return_header=False, return_segments=False):
+    """
+    This function creates the L-band lyot stop used with the vortex charge 2 coronagraph that is installed on Keck 2 
+    that is avaliable for science with NIRC2. The values were extract from drawings of the mask
+    and represent that mask installed around 2018 that is currently avaliable in 2022.
```
[the remark this answers is not in the document]
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

File: hcipy/aperture/realistic.py [hunk 7]
```diff
@@ -1325,3 +1325,167 @@ def func(grid):
...
+    Returns
+    -------
+      	lyot_stop : Field generator
+		A field generator for the Lyot stop.
+
+    """
+    conversion=(10.95/(2*12.05/1000))
+    pupil_diameter = 10.95 #m actual circumscribed diameter
+    actual_segment_flat_diameter = np.sqrt(3)/2 *0.00337*conversion #m actual segment flat-to-flat diameter
+    # iris_ao_segment = np.sqrt(3)/2 * .7 mm (~.606 mm)
```
[the remark this answers is not in the document]
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Deleted

File: hcipy/aperture/realistic.py (hunk 1 above)
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

File: hcipy/aperture/realistic.py (hunk 2 above)
Comment (ivalaginja) [reply hcipy/aperture/realistic.py]: Done

**Comment by ivalaginja** [comment]
>  - [ ]  Put a PNG of the baseline aperture (as checked by the Pytest for an easy check for us and documentation).

@ehpor do you really mean a PNG with this or the `fits.gz` files in the baseline folder of the tests? I wasn't able to find other PNG references.

Sorry for the billion commits and tests, I am flying a little blind since this branch is on a fork and I didn't want to install the fork in dev mode too.

I should be almost done with things, the tests pass when I copy-paste and run them locally. As for the tutorial notebook, I added what I think needs to be there but because of not importing the forked package I wasn't able to run it.

Happy to fix things if there is still something open or something went wrong.

**Comment by ivalaginja** [comment]
Two out of four new tests still failing, let me look into it...

**Comment by ivalaginja** [comment]
@ehpor please advise: the two tests that are failing do so because the reconstruction of the aperture from the segments does not have the central obscuration, while the original aperture does when `with_spiders` is True. This seems to be because Keck, as opposed to LUVOIR A, TMT and ELT does not generate the central obscuration by getting rid of a subset of segments in the center, instead it just uses a circular aperture.

This also made me notice how the Keck aperture uses the `with_spiders` keyword to control the presence of the spiders and central obscuration at the same time, which is not what the other telescopes do. There, the central obscuration is always there. Even when I change this so that the central obscuration is not tethered to that boolean and the obscuration is always there, the tests fail (now four of them instead of only two) because the from-segment reconstruction is not adding the central obscuration.

The simplest solution I see is to (a) always include the central obscuration in Keck, independent of `with_spiders` and (b) add the central obscuration to the reconstruction from segments in the test.

Thoughts?

**Comment by ehpor** [comment]
@ivalaginja With the PNG, I literally mean a screenshot of DS9 of the fits file that serves as the baseline. In the past, even a glance at that file showed that the aperture was completely wrong. And people just let the code generate an aperture and assume that because the test passes, that they've implemented the Keck or VLT aperture correctly.

**Comment by ehpor** [comment]
> @ehpor please advise: the two tests that are failing do so because the reconstruction of the aperture from the segments does not have the central obscuration, while the original aperture does when `with_spiders` is True. This seems to be because Keck, as opposed to LUVOIR A, TMT and ELT does not generate the central obscuration by getting rid of a subset of segments in the center, instead it just uses a circular aperture.
> 
> This also made me notice how the Keck aperture uses the `with_spiders` keyword to control the presence of the spiders and central obscuration at the same time, which is not what the other telescopes do. There, the central obscuration is always there. Even when I change this so that the central obscuration is not tethered to that boolean and the obscuration is always there, the tests fail (now four of them instead of only two) because the from-segment reconstruction is not adding the central obscuration.
> 
> The simplest solution I see is to (a) always include the central obscuration in Keck, independent of `with_spiders` and (b) add the central obscuration to the reconstruction from segments in the test.
> 
> Thoughts?
> 
> Edit: I just realized the solution I suggested above is not really possible because it would require the function `check_aperture()` to be able to identify the case in which it's testing case, and we'd need to put the diameter of the central obscuration in there. I'll stand by until I hear from you Emiel.

The segments that are returned should be the illuminated part of the segment, not the full segment itself. Therefore, the segments in the first ring of Keck should contain the central obscuration. In principle, the sum of all the segments should be equal to the returned segments. This is what the test is checking for. This would be a few extra lines at the end of the function:

```python
if return_segments:
    def segment_with_central_obscuration(segment):
        return lambda grid: segment(grid) * (1 - make_circular_aperture(central_obscuration_diameter)(grid))
    segments = [segment_with_central_obscuration(segment) for segment in segments]
```

or something like this without typos.

**Comment by ivalaginja** [comment]
> In `tests/baseline_for_apertures/keck/*`. They're added by this PR.

@ehpor just to clarify: this is supposed to be a plain png file, uncompressed? Any preferences on the file name? `pupil.png`, `keck.png`, `keck_pupil.png`? Am I seeing right that none of the other folders in `baseline_for_apertures` contains such a png file yet?

**Comment by ivalaginja** [comment]
I don't understand. Everything passes locally and I have no uncommitted changes except for the png file and the .DS_Store files, which is completely irrelevant.

**Comment by ivalaginja** [comment]
Ah. The old reference files do not have the central obscuration in some cases, while what I compare to locally does.

**Comment by ivalaginja** [comment]
There we go.

**Comment by ehpor** [comment]
Thanks for the contribution @vkooten! 🎉

**Pull request merged**